juju/model.py

```diff
@@ -1416,33 +1419,78 @@ async def deploy(
             raise NotImplementedError("trusted is not supported on model version {}".format(self.info.agent_version))
 
-        entity_url = str(entity_url)  # allow for pathlib.Path objects
-        entity_path = Path(entity_url.replace('local:', ''))
-        bundle_path = entity_path / 'bundle.yaml'
-        metadata_path = entity_path / 'metadata.yaml'
+        # Attempt to resolve a charm or bundle based on the URL.
+        # In an ideal world this should be moved to the controller, and we
+        # wouldn't have to deal with this at all.
+        is_local = False
+        is_bundle = False
+        identifier = None
+        origin = None
+        result = None
+
+        # Ensure what we pass in, is a string.
+        entity_url = str(entity_url)
+        if is_local_charm(entity_url) and not entity_url.startswith("local:"):
+            entity_url = "local:{}".format(entity_url)
+        url = URL.parse(str(entity_url))
+        architecture = await self._resolve_architecture(url)
+
+        if Schema.LOCAL.matches(url.schema):
+            entity_url = url.path()
+            entity_path = Path(entity_url)
+            bundle_path = entity_path / 'bundle.yaml'
+
+            identifier = entity_url
+            origin = client.CharmOrigin(source="local", architecture=architecture)
+            if not (entity_path.is_dir() or entity_path.is_file()):
+                raise JujuError('{} path not found'.format(entity_url))
+
+            is_local = True
+            is_bundle = (
+                (entity_url.endswith(".yaml") and entity_path.exists()) or
+                bundle_path.exists()
+            )
 
-        is_local = (
-            entity_url.startswith('local:') or
-            entity_path.is_dir() or
-            entity_path.is_file()
-        )
-        if is_local:
-            entity_id = entity_url.replace('local:', '')
-        else:
-            entity = await self.charmstore.entity(entity_url, channel=channel,
+        elif Schema.CHARM_STORE.matches(url.schema):
+            result = await self.charmstore.entity(str(url),
+                                                  channel=channel,
                                                   include_stats=False)
-            entity_id = entity['Id']
-
-        client_facade = client.ClientFacade.from_connection(self.connection())
-
-        is_bundle = ((is_local and
-                      (entity_id.endswith('.yaml') and entity_path.exists()) or
-                      bundle_path.exists()) or
-                     (not is_local and 'bundle/' in entity_id))
+            identifier = result['Id']
+            origin = client.CharmOrigin(source="charm-store",
+                                        architecture=architecture,
+                                        risk=channel)
+            is_bundle = url.series == "bundle"
+            if not series:
+                series = self._get_series(entity_url, result)
+
+        elif Schema.CHARM_HUB.matches(url.schema):
```

Comment:
All these special cases for local, charm store, and charm hub charms feel bad to me.

I don't think that it's necessarily in the scope of this PR to fix it. But I'd rather see this properly designed. Like there's a nice little class that knows how to do special processing if the charm is local or charm store, and the logic for the bundle just throws stuff at that class, and gets consistent objects back.